<b>Adı, Soyadı =</b> Sena Nur Özdemir <br/>
<b>Numarası =</b> 21040301013<br/>
<b>Dersi =</b> Makine Öğrenmesi FET340<br/>
<b>Öğretim Elemanı =</b> Gökalp Tulum<br/>
<b>Üniversitesi =</b> İstanbul Topkapı Üniversitesi<br/>
<b>Fakültesi =</b> Mühendislik Fakültesi<br/>
<b>Bölümü =</b> Yazılım Mühendisliği<br/>
<b>Sınıfı = </b>3. sınıf

In [158]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sn
from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.naive_bayes import GaussianNB
from sklearn.metrics import confusion_matrix, classification_report, roc_auc_score
from sklearn.neighbors import KNeighborsClassifier
from sklearn.neural_network import MLPClassifier
from sklearn.svm import SVC

Verisetini yükleme ve inceleme işlemleri

In [159]:
df=pd.read_csv('diabetes.csv')
df.head()

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0            6      148             72             35        0  33.6   
1            1       85             66             29        0  26.6   
2            8      183             64              0        0  23.3   
3            1       89             66             23       94  28.1   
4            0      137             40             35      168  43.1   

   DiabetesPedigreeFunction  Age  Outcome  
0                     0.627   50        1  
1                     0.351   31        0  
2                     0.672   32        1  
3                     0.167   21        0  
4                     2.288   33        1

In [160]:
df.isnull().sum()

Pregnancies                 0
Glucose                     0
BloodPressure               0
SkinThickness               0
Insulin                     0
BMI                         0
DiabetesPedigreeFunction    0
Age                         0
Outcome                     0
dtype: int64

In [161]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 768 entries, 0 to 767
Data columns (total 9 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Pregnancies               768 non-null    int64  
 1   Glucose                   768 non-null    int64  
 2   BloodPressure             768 non-null    int64  
 3   SkinThickness             768 non-null    int64  
 4   Insulin                   768 non-null    int64  
 5   BMI                       768 non-null    float64
 6   DiabetesPedigreeFunction  768 non-null    float64
 7   Age                       768 non-null    int64  
 8   Outcome                   768 non-null    int64  
dtypes: float64(2), int64(7)
memory usage: 54.1 KB


In [162]:
diabetSayisi=df['Outcome'].value_counts()
diabetSayisi

Outcome
0    500
1    268
Name: count, dtype: int64

In [163]:
df.describe()

Pregnancies     Glucose  BloodPressure  SkinThickness     Insulin  \
count   768.000000  768.000000     768.000000     768.000000  768.000000   
mean      3.845052  120.894531      69.105469      20.536458   79.799479   
std       3.369578   31.972618      19.355807      15.952218  115.244002   
min       0.000000    0.000000       0.000000       0.000000    0.000000   
25%       1.000000   99.000000      62.000000       0.000000    0.000000   
50%       3.000000  117.000000      72.000000      23.000000   30.500000   
75%       6.000000  140.250000      80.000000      32.000000  127.250000   
max      17.000000  199.000000     122.000000      99.000000  846.000000   

              BMI  DiabetesPedigreeFunction         Age     Outcome  
count  768.000000                768.000000  768.000000  768.000000  
mean    31.992578                  0.471876   33.240885    0.348958  
std      7.884160                  0.331329   11.760232    0.476951  
min      0.000000                  0.078000   21.000000    0.000000  
25%     27.300000                  0.243750   24.000000    0.000000  
50%     32.000000                  0.372500   29.000000    0.000000  
75%     36.600000                  0.626250   41.000000    1.000000  
max     67.100000                  2.420000   81.000000    1.000000

In [164]:
from sklearn.preprocessing import MinMaxScaler
min_max = MinMaxScaler()
df = pd.DataFrame(min_max.fit_transform(df), columns=df.columns)

In [165]:
df.head()

Pregnancies   Glucose  BloodPressure  SkinThickness   Insulin       BMI  \
0     0.352941  0.743719       0.590164       0.353535  0.000000  0.500745   
1     0.058824  0.427136       0.540984       0.292929  0.000000  0.396423   
2     0.470588  0.919598       0.524590       0.000000  0.000000  0.347243   
3     0.058824  0.447236       0.540984       0.232323  0.111111  0.418778   
4     0.000000  0.688442       0.327869       0.353535  0.198582  0.642325   

   DiabetesPedigreeFunction       Age  Outcome  
0                  0.234415  0.483333      1.0  
1                  0.116567  0.166667      0.0  
2                  0.253629  0.183333      1.0  
3                  0.038002  0.000000      0.0  
4                  0.943638  0.200000      1.0

In [166]:
x = df.drop('Outcome', axis=1)
y = df['Outcome']

#verisetini %70 eğitim(train), %30 test olarak bölme
X_train, X_test, y_train, y_test = train_test_split(x, y, test_size=0.3, random_state=42)

<b>Naive Bayes Sınıflandırıcısı</b>

In [167]:
nb = GaussianNB()
nb.fit(X_train, y_train)

y_pred = nb.predict(X_test)

print("Confusion Matrix", confusion_matrix(y_test, y_pred))
print(classification_report(y_test, y_pred))
print("ROC AUC Score:", roc_auc_score(y_test, y_pred))

Confusion Matrix [[119  32]
 [ 27  53]]
              precision    recall  f1-score   support

         0.0       0.82      0.79      0.80       151
         1.0       0.62      0.66      0.64        80

    accuracy                           0.74       231
   macro avg       0.72      0.73      0.72       231
weighted avg       0.75      0.74      0.75       231

ROC AUC Score: 0.7252897350993377


Accuracy = 0.74<br>
Specificity = recall 0.0 = 0.79<br>
Sensitivity = recall 1.0 = 0.66<br>
F1-score 1.0 = 0.62

Confusion Matrix'e bakıldığında, modelin negatif sınıfı (0) 119 doğru ve 32 yanlış, pozitif sınıfı (1) ise 27 yanlış ve 53 doğru sınıflandırılmış. Bu da modelin negatif sınıfı tanımlamada daha başarılı olduğunu, ancak pozitif sınıfı tanımlamada iyileştirme gerektiğini gösteriyor.<br>
Naive Bayes sınıflandırıcısı %74 doğruluk oranıyla ortalamanın üzerinde bir performans sergilemekte. Negatif sınıfı (sınıf 0) tanımada daha başarılı olup bu sınıfta daha yüksek precision ve recall değerlerine sahip, pozitif sınıfı (sınıf 1) tanımada performans biraz daha düşük bu da F1-skoruna yansıyor. F1-skorları negatif sınıfları pozitif sınıflara göre daha iyi tanımladığını gösteriyor bu sebeplede pozitif sınıflar iyileştirilmeli. ROC AUC skoru 0.725 modelin sınıfları ayırt etme yeteneği kabul edilebilir. Genel olarak model iyi bir performans sergiliyor ancak pozitif sınıfları tanımada iyileştirme yapılabilir.

<b>K-En Yakın Komşuluk Sınıflandırıcısı (K-Neighbors)</b>

In [168]:
param_grid = {'n_neighbors': range(1, 31)}

grid_search = GridSearchCV(KNeighborsClassifier(), param_grid, cv=5)
grid_search.fit(X_train, y_train)

best_knn = grid_search.best_estimator_

y_pred_knn = best_knn.predict(X_test)

print("Confusion Matrix\n", confusion_matrix(y_test, y_pred_knn))
print(classification_report(y_test, y_pred_knn))
print("ROC AUC Score:", roc_auc_score(y_test, y_pred_knn))

Confusion Matrix
 [[124  27]
 [ 44  36]]
              precision    recall  f1-score   support

         0.0       0.74      0.82      0.78       151
         1.0       0.57      0.45      0.50        80

    accuracy                           0.69       231
   macro avg       0.65      0.64      0.64       231
weighted avg       0.68      0.69      0.68       231

ROC AUC Score: 0.6355960264900662


Accuracy = 0.69<br>
Specificity = recall 0.0 = 0.82<br>
Sensitivity = recall 1.0 = 0.45<br>
F1-score 1.0 = 0.50

Confusion Matrix'e bakıldığında, negatif sınıfın (0) 124 doğru ve 27 yanlış, pozitif sınıfın (1) ise 44 yanlış ve 36 doğru sınıflandırıldığı görülüyor. Bu da negatif sınıfı daha iyi tanımlamış olsa da pozitif sınıfın tanımlanmasında iyileştirme gerekiyor gibi görünüyor.<br>
Precision, recall ve F1-skorlarına göre, negatif sınıfın (0) performansı daha yüksekken pozitif sınıfın (1) performansı daha düşük. Bu da pozitif sınıfların tanımlanmasında iyileştirme yapılması gerektiğini gösteriyor. ROC AUC skoru 0.64 ise modelin sınıfları ayırt etme yeteneğinin orta derecede olduğunu belirtiyor. Genel olarak K-Neighbors modeli iyi bir performans sergiliyor ancak pozitif sınıfları tanımada iyileştirme yapılması gerekiyor gibi görünüyor.

In [169]:
param_grid = {'n_neighbors': range(1, 51)}
#range değer aralığını değiştirdiğimde çıktımda bir değişiklik gözlemleyemedim

grid_search = GridSearchCV(KNeighborsClassifier(), param_grid, cv=5)
grid_search.fit(X_train, y_train)

best_knn = grid_search.best_estimator_

y_pred_knn = best_knn.predict(X_test)

print("Confusion Matrix\n", confusion_matrix(y_test, y_pred_knn))
print(classification_report(y_test, y_pred_knn))
print("ROC AUC Score:", roc_auc_score(y_test, y_pred_knn))

Confusion Matrix
 [[124  27]
 [ 44  36]]
              precision    recall  f1-score   support

         0.0       0.74      0.82      0.78       151
         1.0       0.57      0.45      0.50        80

    accuracy                           0.69       231
   macro avg       0.65      0.64      0.64       231
weighted avg       0.68      0.69      0.68       231

ROC AUC Score: 0.6355960264900662


In [170]:
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import StandardScaler

#parametrelere bağlı sonuçların biraz daha iyileştiği gözlemlenebilir. bu sonuçları kullanacağım
pipeline = Pipeline([
    ('scaler', StandardScaler()), #özelliklerin ortalamasını 0 ve standart sapmasını 1 olarak ölçeklendirme
    ('knn', KNeighborsClassifier())
])

param_grid = {
    'knn__n_neighbors': range(1, 31),
}

grid_search = GridSearchCV(pipeline, param_grid, cv=5)
grid_search.fit(X_train, y_train)

best_knn = grid_search.best_estimator_

y_pred_knn = best_knn.predict(X_test)

print("Confusion Matrix\n", confusion_matrix(y_test, y_pred_knn))
print(classification_report(y_test, y_pred_knn))
print("ROC AUC Score:", roc_auc_score(y_test, y_pred_knn))

Confusion Matrix
 [[121  30]
 [ 40  40]]
              precision    recall  f1-score   support

         0.0       0.75      0.80      0.78       151
         1.0       0.57      0.50      0.53        80

    accuracy                           0.70       231
   macro avg       0.66      0.65      0.65       231
weighted avg       0.69      0.70      0.69       231

ROC AUC Score: 0.6506622516556292


Accuracy = 0.70<br>
Specificity = recall 0.0 = 0.80<br>
Sensitivity = recall 1.0 = 0.50<br>
F1-score 1.0 = 0.53

<b>MLP(Multi-Layer Perceptron) Sınıflandırıcısı</b>

In [171]:
mlp = MLPClassifier(max_iter=1000, random_state=42)
mlp.fit(X_train, y_train)
y_pred_mlp = mlp.predict(X_test)

print("Confusion Matrix\n", confusion_matrix(y_test, y_pred_mlp))
print(classification_report(y_test, y_pred_mlp))
print("ROC AUC Skoru:", roc_auc_score(y_test, y_pred_mlp))

Confusion Matrix
 [[121  30]
 [ 31  49]]
              precision    recall  f1-score   support

         0.0       0.80      0.80      0.80       151
         1.0       0.62      0.61      0.62        80

    accuracy                           0.74       231
   macro avg       0.71      0.71      0.71       231
weighted avg       0.74      0.74      0.74       231

ROC AUC Skoru: 0.7069122516556292


C:\Users\senan\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\sklearn\neural_network\_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (1000) reached and the optimization hasn't converged yet.
  warnings.warn(


Accuracy = 0.74<br>
Specificity = recall 0.0 = 0.80<br>
Sensitivity = recall 1.0 = 0.61<br>
F1-score 1.0 = 0.62

Confusion Matrix'e bakıldığında negatif sınıfın (0) 121 doğru ve 30 yanlış, pozitif sınıfın (1) ise 31 yanlış ve 49 doğru sınıflandırıldığı görülmektedir. Negatif sınıfı tanımlamada biraz daha başarılı olsa da, pozitif sınıfın tanımlanmasında iyileştirme gerekir.<br>
Precision, recall ve F1-skorlarına göre, negatif sınıfın performansı pozitif sınıfa göre biraz daha yüksek çıkmıştır. Bu da modelin negatif sınıfları daha iyi tanımladığını ancak pozitif sınıfları daha az başarıyla tanımladığını gösteriyor. ROC AUC skoru ise 0.71 olması ise bu da modelin sınıfları ayırt etme yeteneğinin kabul edilebilir olduğunu gösteriyor. Genel olarak MLP modeli iyi bir performans sergilemektedir ancak pozitif sınıfları tanımlamada daha fazla iyileştirme yapılması gerekmekte.

<b>SVM(Support Vector Machines) Sınıflandırıcısı</b>

In [172]:
svm = SVC(probability=True, random_state=42)
svm.fit(X_train, y_train)
y_pred_svm = svm.predict(X_test)

print("Confusion Matrix\n", confusion_matrix(y_test, y_pred_svm))
print(classification_report(y_test, y_pred_svm))
print("ROC AUC Skoru:", roc_auc_score(y_test, y_pred_svm))

Confusion Matrix
 [[126  25]
 [ 32  48]]
              precision    recall  f1-score   support

         0.0       0.80      0.83      0.82       151
         1.0       0.66      0.60      0.63        80

    accuracy                           0.75       231
   macro avg       0.73      0.72      0.72       231
weighted avg       0.75      0.75      0.75       231

ROC AUC Skoru: 0.7172185430463577


Accuracy = 0.74<br>
Specificity = recall 0.0 = 0.83<br>
Sensitivity = recall 1.0 = 0.60<br>
F1-score 1.0 = 0.62

onfusion Matrix'e bakıldığında, negatif sınıfın (0) 126 doğru ve 25 yanlış, pozitif sınıfın (1) ise 32 yanlış ve 48 doğru sınıflandırıldığı gözlemlenmektedir. Model, negatif sınıfı tanımlamada biraz daha başarılı olsa da, pozitif sınıfın tanımlanmasında iyileştirme gerekmektedir.<br>
Precision, recall ve F1-skorlarına göre, negatif sınıfın performansı pozitif sınıfa göre biraz daha yüksek çıkmış. Bu da modelin negatif sınıfları daha iyi tanımladığını, ancak pozitif sınıfları daha az başarıyla tanımladığını gösteriyor. ROC AUC skoru ise 0.72, bu da modelin sınıfları ayırt etme yeteneğinin kabul edilebilir olduğunu gösteriyor. Genel olarak, SVM modeli iyi bir performans sergiliyor ancak pozitif sınıfları tanımlamada daha fazla iyileştirme yapılması gerekmekte.

//FİNAL<br><br>

Naive Bayes<br>
K-Neighbors<br>
MLP<br>
SVM<br><br>


Naive Bayes<br>
########Confusion Matrix#############<br>
[[119  32]<br>
 [ 27  53]]<br>
########sensitivity/specificity/accuracy/f1-score#############<br>
sensitivity =  0.45 <br>
specificity =  0.82 <br>
accuracy =  0.69 <br>
f1-score =  0.50<br><br>


K-Neighbors <br>
########Confusion Matrix#############<br>
 [[121  30]<br>
 [ 40  40]]<br>
########sensitivity/specificity/accuracy/f1-score#############<br>
sensitivity =  0.50 <br>
specificity =  0.80 <br>
accuracy =  0.70 <br>
f1-score =  0.53<br><br>


MLP<br>
########Confusion Matrix#############<br>
 [[121  30]<br>
 [ 31  49]]<br>
########sensitivity/specificity/accuracy/f1-score#############<br>
sensitivity =  0.61 <br>
specificity =  0.80 <br>
accuracy =  0.74 <br>
f1-score =  0.62<br><br>


SVM<br>
########Confusion Matrix#############<br>
 [[126  25]<br>
 [ 32  48]]<br>
########sensitivity/specificity/accuracy/f1-score#############<br>
sensitivity =  0.60 <br>
specificity =  0.83 <br>
accuracy =  0.74 <br>
f1-score =  0.62<br><br><br>




//VİZE<br><br>

LinearRegression<br>
LogisticRegression<br>
DecisionTreeClassifier<br>
Naive Bayes<br><br>

Lineer regression<br>
########Confusion Matrix#############<br>
[[121  30]<br>
 [ 32  48]]<br>
########sensitivity/specificity/accuracy/f1-score#############<br>
sensitivity =  0.6 <br>
specificity =  0.6153846153846154 <br>
accuracy =  0.7316017316017316 <br>
f1-score =  0.6075949367088608<br><br>


Multinominal Lojistik Regresyon <br>
########Confusion Matrix#############<br>
[[124  27]<br>
 [ 34  46]]<br>
########sensitivity/specificity/accuracy/f1-score#############<br>
sensitivity =  0.575 <br>
specificity =  0.6301369863013698 <br>
accuracy =  0.7359307359307359 <br>
f1-score =  0.6013071895424835<br><br>


Karar Ağaçları<br>
########Confusion Matrix#############<br>
[[118  33]<br>
 [ 35  45]]<br>
########sensitivity/specificity/accuracy/f1-score#############<br>
sensitivity =  0.5625 <br>
specificity =  0.5769230769230769 <br>
accuracy =  0.7056277056277056 <br>
f1-score =  0.5696202531645569<br><br>


Bayes<br>
########Confusion Matrix#############<br>
[[119  32]<br>
 [ 27  53]]<br>
########sensitivity/specificity/accuracy/f1-score#############<br>
sensitivity =  0.6625 <br>
specificity =  0.6235294117647059 <br>
accuracy =  0.7445887445887446 <br>
f1-score =  0.6424242424242423

Bütün sonuçlar incelendiğinde SVM sınıflandırıcının sonuçları diğerlerinden daha iyi şekilde değerlendirilebilir. ama bu parametrelerden hangisinin öncelik verildiğine göre değişkenlik göstermektedir.